### ISSUE TITLE
Mismatch between the paper's formula and the code implementation

### ISSUE THREAD

**Issue by alialhousseini** [icannos/glimpse-mds#2]
Hello,
I want to ask about the formula used in the paper for the literal listener which is as follows:
$L_0(d|s) = \frac{LM(s|d)}{\sum_{d \in O} LM(s|d)}$

While in the code, specifically the script `rsasumm/rsa_reranker.py`
The following formula is used:
```python
    @cache
    def S(self, t):
        if t == 0:
            return self.initial_speaker_probas
        else:
            listener = self.L(t - 1)
            # + self.initial_speaker_probas.sum(0, keepdim=True)
            prod = listener * self.rationality
            # Higher rationality focuses on selecting the most relevant candidates.
            return torch.log_softmax(prod, dim=-1)

    @cache
    def L(self, t):
        speaker = self.S(t)
        return torch.log_softmax(speaker, dim=-2)

    def mk_listener_dataframe(self, t):
        self.initial_speaker_probas = self.likelihood_matrix()

        initial_listener_probas = self.L(0)
        ...
```
which in other words will result to a different equation that is:
$L_0 = log(softmax( LM(s|d) )) = log( \frac{e^{LM_{i}}}{\sum_j e^{LM_{j}}})$

Any clarification regarding this confusion is appreciated.
Thanks.

**Comment by icannos** [comment]
Hey,

It is indeed a discrepency between the paper and the implementation, it felt more appropriate to work in log proba / softmax normalization in practice as it's more numerically stable. 

I think you can replace the log_softmax by a `torch.norm(.., dim=-2)` to get closer to the math in the paper. You might be at risk of unstable numerical results if you work with a lot of objects/properties. (The implementation of the logsoftmax avoids these pitfalls).

**Comment by alialhousseini** [comment]
Great! 
Also regarding (uniqueness score / consensuality score), I think the disparency between the paper and the implementation is for the same reason right?
```python
        consensuality_scores = (
            torch.exp(self.L(t))
            * (self.L(t) - torch.log(uniform_distribution_over_source_texts))
        ).sum(0).cpu().numpy()
```
Thank you for this clarification.

**Comment by icannos** [comment]
Actually this is exactly what is in the paper I think, just taking into account that in the code L and S corresponds to the log probs.

$KL(L(. | p ) || U) = \sum_o L(o | p) log [  L(o | p)  /  U] $
